lapack/gonum/dlatbs.go

```diff
@@ -0,0 +1,449 @@
+			// Initially, G_0 = max{x(i), i=1,...,n}.
+			grow = 1 / math.Max(xBnd, smlnum)
+			xBnd = grow
+			for j := jFirst; j != jLast; j += jInc {
+				if grow <= smlnum {
+					// Exit the loop because the growth factor is too small.
+					goto skipComputeGrow
+				}
+				// M_j = G_{j-1} / abs(A[j,j])
+				tjj := math.Abs(ab[j*ldab+maind])
+				xBnd = math.Min(xBnd, math.Min(1, tjj)*grow)
+				if tjj+cnorm[j] >= smlnum {
+					// G_j = G_{j-1}*( 1 + cnorm[j] / abs(A[j,j]) )
+					grow *= tjj / (tjj + cnorm[j])
+				} else {
+					// G_j could overflow, set grow to 0.
+					grow = 0
+				}
+			}
+			grow = xBnd
+		} else {
+			// A is unit triangular.
+			//
+			// Compute grow = 1/G_j, where G_0 = max{x(i), i=1,...,n}.
+			grow = math.Min(1, 1/math.Max(xBnd, smlnum))
+			for j := jFirst; j != jLast; j += jInc {
+				if grow <= smlnum {
+					// Exit the loop because the growth factor is too small.
+					goto skipComputeGrow
+				}
+				// G_j = G_{j-1}*( 1 + cnorm[j] )
+				grow /= 1 + cnorm[j]
+			}
+		}
+	} else {
+		// Compute the growth in A^T * x = b.
+		if diag == blas.NonUnit {
+			// A is non-unit triangular.
+			//
+			// Compute grow = 1/G_j and xBnd = 1/M_j.
+			// Initially, G_0 = max{x(i), i=1,...,n}.
+			grow = 1 / math.Max(xBnd, smlnum)
+			xBnd = grow
+			for j := jFirst; j != jLast; j += jInc {
+				if grow <= smlnum {
+					// Exit the loop because the growth factor is too small.
+					goto skipComputeGrow
+				}
+				// G_j = max( G_{j-1}, M_{j-1}*( 1 + cnorm[j] ) )
+				xj := 1 + cnorm[j]
+				grow = math.Min(grow, xBnd/xj)
+				// M_j = M_{j-1}*( 1 + cnorm[j] ) / abs(A[j,j])
+				tjj := math.Abs(ab[j*ldab+maind])
+				if xj > tjj {
+					xBnd *= tjj / xj
+				}
+			}
+			grow = math.Min(grow, xBnd)
+		} else {
+			// A is unit triangular.
+			//
+			// Compute grow = 1/G_j, where G_0 = max{x(i), i=1,...,n}.
+			grow = math.Min(1, 1/math.Max(xBnd, smlnum))
+			for j := jFirst; j != jLast; j += jInc {
+				if grow <= smlnum {
+					// Exit the loop because the growth factor is too small.
+					goto skipComputeGrow
+				}
+				// G_j = G_{j-1}*( 1 + cnorm[j] )
+				grow /= 1 + cnorm[j]
+			}
+		}
+	}
+skipComputeGrow:
+
+	if grow*tscal > smlnum {
+		// The reciprocal of the bound on elements of X is not too small, use
+		// the Level 2 BLAS solve.
+		bi.Dtbsv(uplo, trans, diag, n, kd, ab, ldab, x, 1)
+		// Scale the column norms by 1/tscal for return.
+		if tscal != 1 {
+			bi.Dscal(n, 1/tscal, cnorm, 1)
+		}
+		return 1
+	}
+
+	// Use a Level 1 BLAS solve, scaling intermediate results.
+
+	scale = 1
+	if xMax > bignum {
+		// Scale x so that its components are less than or equal to bignum in
+		// absolute value.
+		scale = bignum / xMax
+		bi.Dscal(n, scale, x, 1)
+		xMax = bignum
+	}
+
+	if noTran {
+		// Solve A * x = b.
+		for j := jFirst; j != jLast; j += jInc {
+			// Compute x[j] = b[j] / A[j,j], scaling x if necessary.
+			xj := math.Abs(x[j])
+			tjjs := tscal
+			if diag == blas.NonUnit {
```

Comment:
I think I got lost here. There is a condition in the Fortran for `tscal.EQ.one` dropping through to the scaling of `x` that doesn't seem to happen here.

Reply:
Yes, this one is a bit tricky. I omitted the `goto` because: if `tscal` is 1, then `tjjs` and `tjj` are 1 and so the first `case` of the `switch` will trigger. In that `case` the `if` branch will not be taken, so essentially what the `goto` is skipping is only the division by `tjjs` and the absolute value. I thought that doing these two operations is a fair price for a clearer code flow because for example the declaration of tjj would have to happen before the goto and tjj would be initialized from tjjs only conditionally.

Reply:
Thanks